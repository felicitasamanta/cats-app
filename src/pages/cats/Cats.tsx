import { useEffect } from "react";
import { Cat } from "./common/components/Cat";
import classes from "./common/styles/Cats.module.css";
import { LIMIT, useCats } from "./common/hooks/useCats";
import { Filters } from "./common/components/Filters";
import { useQueryParams } from "@/common/hooks/useQueryParams";
import { LoaderContainer } from "@/common/components/LoaderContainer";
import { Pagination } from "@/common/components/Pagination";

const Cats = () => {
  const { params } = useQueryParams();
  const { isLoading, data, total, refetch } = useCats();
  const { order, page, breed_ids: breedId, image_id: imageId } = params;

  useEffect(() => {
    refetch();
  }, [order, page, breedId, imageId]);

  return (
    <div>
      <Filters isLoading={isLoading} />
      <LoaderContainer isLoading={isLoading}>
        <ul className={classes.list} data-testid="cats">
          {data?.map((cat) => (
            <Cat key={cat.id} cat={cat} />
          ))}
        </ul>
        <Pagination total={total as number} limit={LIMIT} />
      </LoaderContainer>
    </div>
  );
};

export default Cats;
