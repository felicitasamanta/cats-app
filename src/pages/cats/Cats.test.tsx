import { TestsWrapper } from "@/common/tests/TestsWrapper";
import {
  cleanup,
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { afterAll, afterEach, beforeAll, describe, expect, it } from "vitest";
import { setupServer } from "msw/node";
import { rest } from "msw";
import Cats from "./Cats";

// https://api.thecatapi.com/v1/images/search?limit=10&&api_key=live_23gBCX5oksoKw4hl4o6DMzmjuh2APvDNxWAAC2Ctrg9zgEcomnY44ce7mSbfuyjF
const cats = [
  {
    breeds: [],
    id: "4s4",
    url: "https://cdn2.thecatapi.com/images/4s4.jpg",
    width: 500,
    height: 332,
  },
  {
    breeds: [],
    id: "77t",
    url: "https://cdn2.thecatapi.com/images/77t.jpg",
    width: 500,
    height: 375,
  },
  {
    breeds: [],
    id: "ch8",
    url: "https://cdn2.thecatapi.com/images/ch8.jpg",
    width: 420,
    height: 583,
  },
];

const server = setupServer(
  rest.get("https://api.thecatapi.com/v1/breeds", (_, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json([
        {
          id: "abys",
          name: "Abyssinian",
        },
        {
          id: "aege",
          name: "Aegean",
        },
      ])
    );
  }),
  rest.get("https://api.thecatapi.com/v1/images/search", (_, res, ctx) => {
    return res(ctx.status(200), ctx.json(cats));
  })
);

describe("Cats", () => {
  beforeAll(() => {
    server.listen({ onUnhandledRequest: "error" });
  });

  afterEach(() => {
    cleanup();
    server.resetHandlers();
  });

  afterAll(() => {
    server.close();
  });

  it("should match snapshot", async () => {
    const component = render(<Cats />, { wrapper: TestsWrapper });

    await waitForElementToBeRemoved(() => screen.getByTestId("loader"));
    expect(component).toMatchSnapshot();
  });

  it("should render a list of cats", async () => {
    render(<Cats />, { wrapper: TestsWrapper });

    await waitForElementToBeRemoved(() => screen.getByTestId("loader"));
    const list = screen.getAllByTestId("cat");
    expect(list.length).toEqual(3);
    expect(list[0].getElementsByTagName("img").item(0)?.src).toEqual(
      cats[0].url
    );
    expect(list[1].getElementsByTagName("img").item(0)?.src).toEqual(
      cats[1].url
    );
    expect(list[2].getElementsByTagName("img").item(0)?.src).toEqual(
      cats[2].url
    );
  });
});
