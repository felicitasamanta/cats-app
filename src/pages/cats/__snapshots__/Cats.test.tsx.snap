// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cats > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div>
          <div
            class="_container_c9f2fc"
          >
            <div
              class="_container_dde768"
            >
              <input
                class="_input_dde768"
                data-testid="search"
                id="search"
                name="search"
                placeholder="Search..."
                value=""
              />
            </div>
            <div
              class="_container_7c5b13"
            >
              <div
                class="_input_7c5b13"
                data-testid="dropdown"
              >
                <div
                  class="_selected_7c5b13"
                  data-testid="dropdown-selected"
                >
                  Random
                </div>
                <div
                  class="_tools_7c5b13"
                >
                  <div
                    class="_tool_7c5b13"
                  >
                    <svg
                      height="20"
                      viewBox="0 0 30 30"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M 2.65625 6.25 L 1.34375 7.75 L 11.34375 16.75 L 12 17.34375 L 12.65625 16.75 L 22.65625 7.75 L 21.34375 6.25 L 12 14.65625 L 2.65625 6.25 z"
                        fill="#000000"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="_container_7c5b13"
            >
              <div
                class="_input_7c5b13"
                data-testid="dropdown"
              >
                <div
                  class="_selected_7c5b13"
                  data-testid="dropdown-selected"
                >
                  All
                </div>
                <div
                  class="_tools_7c5b13"
                >
                  <div
                    class="_tool_7c5b13"
                  >
                    <svg
                      height="20"
                      viewBox="0 0 30 30"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M 2.65625 6.25 L 1.34375 7.75 L 11.34375 16.75 L 12 17.34375 L 12.65625 16.75 L 22.65625 7.75 L 21.34375 6.25 L 12 14.65625 L 2.65625 6.25 z"
                        fill="#000000"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <ul
            class="_list_f48831"
          >
            <li
              class="_container_ceb8ea"
              data-testid="cat"
            >
              <a
                href="/cats/4s4"
              >
                <img
                  alt="Cat 4s4"
                  class="_img_ceb8ea"
                  src="https://cdn2.thecatapi.com/images/4s4.jpg"
                />
              </a>
            </li>
            <li
              class="_container_ceb8ea"
              data-testid="cat"
            >
              <a
                href="/cats/77t"
              >
                <img
                  alt="Cat 77t"
                  class="_img_ceb8ea"
                  src="https://cdn2.thecatapi.com/images/77t.jpg"
                />
              </a>
            </li>
            <li
              class="_container_ceb8ea"
              data-testid="cat"
            >
              <a
                href="/cats/ch8"
              >
                <img
                  alt="Cat ch8"
                  class="_img_ceb8ea"
                  src="https://cdn2.thecatapi.com/images/ch8.jpg"
                />
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div>
        <div
          class="_container_c9f2fc"
        >
          <div
            class="_container_dde768"
          >
            <input
              class="_input_dde768"
              data-testid="search"
              id="search"
              name="search"
              placeholder="Search..."
              value=""
            />
          </div>
          <div
            class="_container_7c5b13"
          >
            <div
              class="_input_7c5b13"
              data-testid="dropdown"
            >
              <div
                class="_selected_7c5b13"
                data-testid="dropdown-selected"
              >
                Random
              </div>
              <div
                class="_tools_7c5b13"
              >
                <div
                  class="_tool_7c5b13"
                >
                  <svg
                    height="20"
                    viewBox="0 0 30 30"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M 2.65625 6.25 L 1.34375 7.75 L 11.34375 16.75 L 12 17.34375 L 12.65625 16.75 L 22.65625 7.75 L 21.34375 6.25 L 12 14.65625 L 2.65625 6.25 z"
                      fill="#000000"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <div
            class="_container_7c5b13"
          >
            <div
              class="_input_7c5b13"
              data-testid="dropdown"
            >
              <div
                class="_selected_7c5b13"
                data-testid="dropdown-selected"
              >
                All
              </div>
              <div
                class="_tools_7c5b13"
              >
                <div
                  class="_tool_7c5b13"
                >
                  <svg
                    height="20"
                    viewBox="0 0 30 30"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M 2.65625 6.25 L 1.34375 7.75 L 11.34375 16.75 L 12 17.34375 L 12.65625 16.75 L 22.65625 7.75 L 21.34375 6.25 L 12 14.65625 L 2.65625 6.25 z"
                      fill="#000000"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <ul
          class="_list_f48831"
        >
          <li
            class="_container_ceb8ea"
            data-testid="cat"
          >
            <a
              href="/cats/4s4"
            >
              <img
                alt="Cat 4s4"
                class="_img_ceb8ea"
                src="https://cdn2.thecatapi.com/images/4s4.jpg"
              />
            </a>
          </li>
          <li
            class="_container_ceb8ea"
            data-testid="cat"
          >
            <a
              href="/cats/77t"
            >
              <img
                alt="Cat 77t"
                class="_img_ceb8ea"
                src="https://cdn2.thecatapi.com/images/77t.jpg"
              />
            </a>
          </li>
          <li
            class="_container_ceb8ea"
            data-testid="cat"
          >
            <a
              href="/cats/ch8"
            >
              <img
                alt="Cat ch8"
                class="_img_ceb8ea"
                src="https://cdn2.thecatapi.com/images/ch8.jpg"
              />
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
